### PR TITLE
fix CNFE when running an app packaged with custom main class

### PR DESCRIPTION
**Pull request by machi1990**
When a user defined the main class via `quarkus.package.main-class=foo.bar.MainClass` property the
following error `Error: Could not find or load main class foo.bar.MainClass. Caused by:
java.lang.ClassNotFoundException: foo.bar.MainClass` occurred when launching the application and app could not be started

/cc @dmlloyd @stuartwdouglas

### CODE REVIEW

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java
Comment (gsmet) [inline integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: Please add another test when you do that.

File: core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java
Comment (gsmet) [inline core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: These classes are unused?

**Review by gsmet — CHANGES_REQUESTED** [review]
Added a couple of comments/questions.

File: core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java
Comment (machi1990) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: Yes, looks so to me. I'd like another opinion though @stuartwdouglas @dmlloyd

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java
Comment (machi1990) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: I think the test in here is enough as without this patch if you add the mainClass property (via build tool e.g maven or in application.properties )you could not launch the application - https://github.com/quarkusio/quarkus/blob/95590681ea0a9c94be8f81a16114e0eca0495ab5/integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java#L44-L47.

File: core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java
Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: I believe that the intent of this class was to provide a class that can be run from an IDE.

Comment (machi1990) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: Oh good to know, thanks David. So we should keep it. Same question for `QuarkusReplacement.java` contains a substitution of the generated main class? Can it be deleted?

Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: I'm not sure what the current status of that is TBH.

Comment (machi1990) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: I’ll try removing it and keep the Substitutee and see if it the substitutions are still needed.

Comment (machi1990) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: I kept the `Quarkus` for IDE runs and removed the `QuarkusReplacement`.

Comment (stuartwdouglas) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: These are leftovers from the early days, they can both go.

Comment (machi1990) [reply core/runtime/src/main/java/io/quarkus/runtime/Quarkus.java]: So the two classes are gone. This is ready for another round of review.

File: integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java
Comment (stuartwdouglas) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: I think this should really be a different test to avoid confusion. I also don't understand how this test passes? Where does foo.bar.MainClass come from?

Comment (machi1990) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: Yes, we can have another test. I wanted to quickly add the mainClass property when packaging in this existing JarRunner test to avoid straining the CI with yet another similar test. 

foo.bar.MainClass is just an arbitrary class name supplied as configuration property to override the default ‘io.quarkus.runner.GeneratedMain’

Comment (stuartwdouglas) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: But the class doesn't exist so why does the test pass?

Comment (machi1990) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: It passes because of my second commit https://github.com/quarkusio/quarkus/pull/5542/commits/a069ecba35fcd5d9cf0f977b3ec9d091b2233562

Comment (machi1990) [reply integration-tests/maven/src/test/java/io/quarkus/maven/it/JarRunnerIT.java]: Is this how the mainClass property should be used? Or it is a user defined one?

**Comment by stuartwdouglas** [comment]
Can we hold this one for now, as this likely ties into what we are going to do with command mode and main methods (which is an ongoing discussion on the list). Hopefully we will have an answer for that soon and we can define the semantics exactly what the main-class attribute is, as at the moment it is not particularly well defined.

**Comment by machi1990** [comment]
> Can we hold this one for now, as this likely ties into what we are going to do with command mode and main methods (which is an ongoing discussion on the list). Hopefully we will have an answer for that soon and we can define the semantics exactly what the main-class attribute is, as at the moment it is not particularly well defined.

Thanks for the feedback @stuartwdouglas. I noticed the issue when I was playing with jib.  

Sure, let's hold this. Can we have a separate PR with only the first commit removing the `Quarkus` and `QuarkusReplacement` classes ?

**Comment by stuartwdouglas** [comment]
Sure, sounds good.

**Comment by machi1990** [comment]
> Sure, sounds good.

Let me prepare it.

**Comment by gsmet** [comment]
I think I will close this one for now. All this will probably be changed soon with the command mode.

**Pull request closed without merging**
